src/audio/frame.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <QDebug>
+#include <cmath>
+#include <limits>
+
+#include "engine/engine.h"
+#include "util/fpclassify.h"
+
+namespace mixxx {
+namespace audio {
+/// FrameDiff_t can be used to store the difference in position between
+/// two frames and to store the length of a segment of track in terms of frames.
+typedef double FrameDiff_t;
+
+/// FramePos defines the position of a frame in a track
+/// with respect to a fixed origin, i.e. start of the track.
+class FramePos final {
+  public:
+    typedef double value_t;
+    static constexpr value_t kStartValue = 0;
+    static constexpr value_t kInvalidValue = std::numeric_limits<FramePos::value_t>::quiet_NaN();
+
+    constexpr FramePos()
+            : m_framePosition(kInvalidValue) {
+    }
+
+    constexpr explicit FramePos(value_t framePosition)
+            : m_framePosition(framePosition) {
+    }
+
+    static constexpr FramePos fromEngineSamplePos(double engineSamplePos) {
+        return FramePos(engineSamplePos / mixxx::kEngineChannelCount);
+    }
+
+    constexpr double toEngineSamplePos() const {
+        return value() * mixxx::kEngineChannelCount;
+    }
+
+    bool isValid() const {
+        return !util_isnan(m_framePosition);
+    }
+
+    void setValue(value_t framePosition) {
+        m_framePosition = framePosition;
+    }
+
+    constexpr value_t value() const {
+        return m_framePosition;
+    }
+
+    bool isFractional() const {
+        value_t integerPart;
+        return std::modf(value(), &integerPart) != 0;
+    }
+
+    [[nodiscard]] FramePos toFullFrames() const {
```

Comment:
Floor is not int cast.

Reply:
When I mentioned the naming is not ideal I had something like `toLowerFrameBoundary()` in mind.

Reply:
@uklotzde Done.

Reply:
I guess that for the future we need a more sophisticated solution:

- If the value is close (< `epsilon`) to a beat then use `std::round`
- Otherwise use either the lower or upper boundary, depending on the use case.

The value of `epsilon` may also vary.